kubernetes/vault-snapshot.sh

```diff
@@ -7,8 +7,22 @@ VAULT_TOKEN=$(vault write -field=token  auth/kubernetes/login role="${VAULT_ROLE
 export VAULT_TOKEN
 
 # create snapshot
-
 vault operator raft snapshot save /vault-snapshots/vault_"$(date +%F-%H%M)".snapshot
 
 # upload to s3
 s3cmd put /vault-snapshots/* "${S3_URI}" --host="${S3_HOST}" --host-bucket="${S3_BUCKET}"
+
+# remove expired snapshots
+if [ "${S3_EXPIRE_DAYS}" ]; then
+    s3cmd ls "${S3_URI}" --host="${S3_HOST}" --host-bucket="${S3_BUCKET}" | while read -r line; do
+        createDate=$(echo "$line" | awk '{print $1" "$2}')
+        createDate=$(date -d"$createDate" +%s)
+        olderThan=$(date --date @$(($(date +%s) - 86400*S3_EXPIRE_DAYS)) +%s)
+        if [ "$createDate" -lt "$olderThan" ]; then
```

Comment:
Honestly seeing date math in shell scripts makes me wonder if we would be better of implementing the whole thing in python and using both python date math, as well as a python library to interact with S3.

Reply:
Sure, anything works for me. Maybe we can branch that off into a different story ("re-implementation in Pyhton" or similar)? For now I'm just glad I have a solution that "works" :smile:

Reply:
@tongpu +1 for Python implementation:
* We don't have to write to a file (we can read the file stream and store the object directly from memory to s3)
* We can get rid of `S3_URI`

I did not have too much time this evening, but this one already works pretty good (snapshot & upload). The iteration through the objects is not done yet (retention part): https://github.com/adfinis/vault-raft-backup-agent/compare/feat/remove-expired-snapshots...feat/python

Todo:
* Kubernetes auth (almost done)
* Iteration through objects for retention purposes
* Make it pretty :sparkles:

Reply:
https://github.com/adfinis/vault-raft-backup-agent/issues/30